### PR TITLE
`ConvertedTableLoader` minor bug fixes

### PR DESCRIPTION
**Pull request by hepcat72**
## Summary Change Description

See in-line comments.

## Affected Issues/Pull Requests

- Resolves no documented issue - just some test catch-up for previously implemented issues
- Merges into PR #1323

## Review Notes

See comments in-line.

## Checklist

This pull request will be merged once the following requirements are met.  The
author and/or reviewers should uncheck any unmet requirements:

- Review requirements
  - Minimum approvals: 1 
  - No changes requested
  - All blocking issues resolved by reviewers
  - Specific reviewers: @__add_username_here__
    
  - Review period: 2 days 
- Associated issue/pull request requirements:
  
  - [x] All requirements in affected issues marked "resolved" are satisfied
  - [x] All required pull requests are merged *(or none)*
- Basic requirements
  
  - [x] [All linters pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting)
  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All conflicts resolved
- Overhead requirements
  
  - [x] [New/changed method tests implemented/updated *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [ ] [Updated *Unreleased* section of `changelog.md` *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md)
  - [x] [Migrations created & committed *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)

### CODE REVIEW

File: DataRepo/loaders/base/converted_table_loader.py [hunk 1]
```diff
@@ -659,16 +659,18 @@ def revert_headers(self, headers):
...
-                        and len(merge_dict["right_all_columns"]) > 0
-                        and rh in merge_dict["right_all_columns"]
+                        merge_dict["right_columns"] is not None
+                        and len(merge_dict["right_columns"]) > 0
+                        and rh in merge_dict["right_columns"]
                     ):
                         rev_headers_dict[merge_dict["right_sheet"]].append(rh)
                     elif merge_dict["next_merge_dict"] is None:
-                        rev_headers_dict[right_sheet].append(rh)
+                        rev_headers_dict[left_sheet].append(rh)
```
Comment (hepcat72) [inline DataRepo/loaders/base/converted_table_loader.py:673]: C. This code needs more improvement.  It is only used for reporting missing headers in an exception, so it's not terribly important.  There's an issue with the `right_all_columns` that I have yet to fix.  It is intended to have user-supplied headers added in, but there's a state issue and always happens to be `None` here, so I switched to `right_columns` instead (which are the default supported columns).  Don't think about it too hard.  This needs rewritten.

File: DataRepo/loaders/base/converted_table_loader.py [hunk 2]
```diff
@@ -957,6 +961,8 @@ def fill_down_nan_columns(self, indf):
             for col in self.nan_filldown_columns:
                 if col in outdf.columns:
                     outdf[col].ffill(inplace=True)
+                    # In case an empty value was left at the top...
+                    outdf[col].bfill(inplace=True)
```
Comment (hepcat72) [inline DataRepo/loaders/base/converted_table_loader.py:965]: C. This is the important bit.  This allows formulas to be backfilled after sorting the dataframe, just in case the first row ends up with an empty formula.  This occurred in one of my other tests in code that's not committed here.  There were so few rows in the test, there was basically a 50/50 chance the first row would have an empty formula, and I'm glad it did.  This is a satisfying bug fix.

File: DataRepo/tests/loaders/base/test_converted_table_loader.py [hunk 3]
```diff
@@ -10,7 +11,7 @@
 
 
 @isolate_apps("DataRepo.tests.apps.loader")
-class TestConvertedLoaderTests(TracebaseTestCase):
+class ConvertedTableLoaderTests(TracebaseTestCase):
```
Comment (hepcat72) [inline DataRepo/tests/loaders/base/test_converted_table_loader.py:14]: C. This test class name wasn't following the convention, so I changed it.

File: DataRepo/tests/loaders/base/test_converted_table_loader.py [hunk 4]
```diff
@@ -76,10 +77,6 @@ class TestTableLoader(ConvertedTableLoader):
                 CORRECTED="Corrected Abundance",
             )
             DataRequiredHeaders = [
-                "MEDMZ",
-                "MEDRT",
-                "ISOTOPELABEL",
-                "FORMULA",
```
Comment (hepcat72) [inline DataRepo/tests/loaders/base/test_converted_table_loader.py:82]: C. This is a test for the `csv` fix for accucor from another PR.  These columns need to be optional if we don't require the `Original` sheet.  This isn't technically a test of tracebase code.  It's an independent test that essentially tests that optional sheets aren't required by the "converted-to" format.

File: DataRepo/tests/loaders/base/test_converted_table_loader.py [hunk 5]
```diff
@@ -160,7 +157,7 @@ class TestConvertedLoader1(TestTableLoader):
                 }
             }
 
-            merge_dict = cls.accucor_merge_dict
+            merge_dict = deepcopy(cls.accucor_merge_dict)
```
Comment (hepcat72) [inline DataRepo/tests/loaders/base/test_converted_table_loader.py:160]: C. I added some copy methods here and there until I eliminated an issue where an inadvertant change to a class attribute persisted through to the next test, making certain tests pass or fail based on what other tests were running.

File: DataRepo/tests/loaders/base/test_converted_table_loader.py [hunk 6]
```diff
@@ -188,13 +185,6 @@ class TestConvertedLoader1(TestTableLoader):
             )
 
             OrigDataRequiredHeaders = {
-                "Original": [
-                    "FORMULA",
-                    "MEDMZ",
-                    "MEDRT",
-                    "ISOTOPELABEL",
-                    "ORIGCOMPOUND",
-                ],
```
Comment (hepcat72) [inline DataRepo/tests/loaders/base/test_converted_table_loader.py:197]: C. Same as above.  The original sheet is optional (accucor).

File: DataRepo/tests/loaders/base/test_converted_table_loader.py [hunk 7]
```diff
@@ -298,11 +288,11 @@ class TestConvertedLoader2(TestTableLoader):
...
-                "formula": str,
-                "medMz": float,
-                "medRt": float,
-                "isotopeLabel": str,
-                "compound": str,
+                "FORMULA": str,
+                "MEDMZ": float,
+                "MEDRT": float,
+                "ISOTOPELABEL": str,
+                "COMPOUND": str,
```
Comment (hepcat72) [inline DataRepo/tests/loaders/base/test_converted_table_loader.py:295]: C. This was an uncaught but.  This class attribute should have had header keys as keys, not header names as keys.

File: DataRepo/tests/loaders/base/test_converted_table_loader.py [hunk 8]
```diff
@@ -347,13 +337,16 @@ class TestConvertedLoader2(TestTableLoader):
...
+    def setUp(self):
+        super().setUp()
+        mdl = self.generate_test_model()
+        TestTableLoader = self.generate_test_table_loader(mdl)
+        self.TestConvertedLoader1 = self.generate_test_converted_loader1(
+            TestTableLoader
+        )
+        self.TestConvertedLoader2 = self.generate_test_converted_loader2(
+            TestTableLoader
+        )
```
Comment (hepcat72) [inline DataRepo/tests/loaders/base/test_converted_table_loader.py:349]: C. Another test-state change to fix a test that passed when run by itself, but failed when run after a test that changed a class attribute.

File: DataRepo/tests/loaders/base/test_converted_table_loader.py [hunk 9]
```diff
@@ -585,6 +579,13 @@ def get_converted_accucor_df(self):
             },
         )
 
+    def get_converted_corrected_accucor_df(self):
+        """This removes all of the columns that come from the original sheet, so it returns a dataframe that represents
+        just the corrected sheet."""
+        return self.get_converted_accucor_df()[
+            ["Compound", "mzXML Name", "Corrected Abundance"]
+        ]
```
Comment (hepcat72) [inline DataRepo/tests/loaders/base/test_converted_table_loader.py:587]: C. Another test about the original sheet being optional.

File: DataRepo/tests/loaders/base/test_converted_table_loader.py [hunk 10]
```diff
@@ -596,17 +597,29 @@ def test_convert_df_accucor_excel(self):
...
+        The only thing that's missing (from a Tracebase perspective) is the formula column, but the loading code in that
+        instance, grabs in from the compound record.  Note, this is just an example, written to test a
+        ConvertedTableLoader feature (skip merging if the 1 required sheet is provided), independent of Tracebase, but
+        this was also written to approximate what would happen in TraceBase."""
+        tcl = self.TestConvertedLoader1(  # pylint: disable=not-callable
+            df=self.ACCUCOR_DF_DICT["Corrected"]
+        )
+        tcl.convert_df()
+        aes = tcl.aggregated_errors_object
+        self.assertEqual(0, len(aes.exceptions))
```
Comment (hepcat72) [inline DataRepo/tests/loaders/base/test_converted_table_loader.py:611]: C. Tests that the original sheet is truly optional.

File: DataRepo/tests/loaders/base/test_converted_table_loader.py [hunk 11]
```diff
@@ -596,17 +597,29 @@ def test_convert_df_accucor_excel(self):
...
+                df=self.ACCUCOR_DF_DICT["Original"].copy(deep=True)
             ).convert_df()
         aes = ar.exception
         self.assertEqual(1, len(aes.exceptions))
         self.assertIsInstance(aes.exceptions[0], RequiredHeadersError)
         self.assertIn(
-            "missing: {'Corrected': ['medMz', 'medRt', 'isotopeLabel', 'formula']}",
+            "missing: {'Corrected': ['Compound']}",
             str(aes.exceptions[0]),
         )
```
Comment (hepcat72) [inline DataRepo/tests/loaders/base/test_converted_table_loader.py:624]: C. Tests that the corrected sheet is required (csv).

File: DataRepo/tests/loaders/base/test_converted_table_loader.py [hunk 12]
```diff
@@ -720,17 +733,11 @@ def test_extra_columns_excluded_accucor_excel(self):
...
-        }
-        expected = self.get_converted_accucor_df()
+        tmpdict = self.CORR_DICT.copy()
+        tmpdict["adductName"] = [0, 0, 0, 0]
+        expected = self.get_converted_corrected_accucor_df()
         outdf = self.TestConvertedLoader1(  # pylint: disable=not-callable
-            df=tmpdict
+            df=pd.DataFrame.from_dict(tmpdict)
         ).convert_df()
         pd.testing.assert_frame_equal(expected, outdf, check_like=True)
```
Comment (hepcat72) [inline DataRepo/tests/loaders/base/test_converted_table_loader.py:742]: C. Another test related to the original sheet being optional.  It was previously not correct and needed to be converted to supplying a dataframe instead of a dict.

File: DataRepo/tests/loaders/base/test_converted_table_loader.py [hunk 13]
```diff
@@ -869,14 +877,46 @@ def test_get_required_headers(self):
...
-                    "medRt",
-                    "isotopeLabel",
-                    "compound",
                     "Compound",
                     "C_Label",
                 ]
             ),
             set(rh),
         )
+
```
Comment (hepcat72) [inline DataRepo/tests/loaders/base/test_converted_table_loader.py:886]: C. New method tests...

File: DataRepo/loaders/base/converted_table_loader.py (hunk 2 above)
Comment (fkang-pu) [reply DataRepo/loaders/base/converted_table_loader.py:965]: In the future, it's better to add a few examples (test cases would be nice) to prove `ffill `and `bfill`  indeed fill in null value(s) as expected.

**Review by fkang-pu — APPROVED** [review]
Nice work for adding those tests in this PR!

File: DataRepo/loaders/base/converted_table_loader.py (hunk 2 above)
Comment (hepcat72) [reply DataRepo/loaders/base/converted_table_loader.py:965]: It is indirectly tested, but I agree.
